GeneticAlgorithm.py

```diff
@@ -0,0 +1,301 @@
+    new_assignment_sample2 = random.sample(Leftover_Facility2.tolist(), len(Leftover_Facility2))
+    l=0 #Counters to keep track of elements in new assignment variable
+    m=0
+    
+    for k in range(len(offspring1)):
+        if(offspring1[k] == -1): #For every location where facilities are not assigned for offspring1
+            offspring1[k] = new_assignment_sample1[l] #Add missing sampled facilities into offspring
+            l = l + 1
+        if(offspring2[k] == -1): #For every location where facilities are not assigned for offspring2
+            offspring2[k] = new_assignment_sample2[m]
+            m = m + 1
+        
+    new_offspring1.append(offspring1)
+    new_offspring1.append(cost)
+    
+    new_offspring2.append(offspring2)
+    new_offspring2.append(cost)
+    
+    return [new_offspring1,new_offspring2]
+
+
+
+
+def selection(X_pop, k):
+    """
+        Returns the best solutions by picking k solutions in the population
+            Parameter:
+                X_pop (list): list of solutions from population
+                k (int): number of selected solutions to be picked before choosing the one best solution
+            Returns:
+                k_selected_solutions[0] (list): Best solution out of the k chosen
+    """
+    k_selected_solutions = random.sample(X_pop,k) #Sample without replacement k samples in the population
+    k_selected_solutions.sort(key = lambda x: x[1]) #Sort the solutions in ascending order by the fitness value
+    return k_selected_solutions[0]
+
+
+
+
+
+
+def genetic_Algorithm(n,N_pop,distance_data, flow_data,MaxIter,k):
+    """
+        The Main function to call to run the Genetic Algorithm. Returns the optimal solution
+            Parameter:
+                n (int): number of facilities or location
+                N_pop (int): initial population size
+                distance_data (matrix): Distance Data matrix
+                flow_data (matrix): Flow Data matrix
+                MaxIter (int): Maximum number of iterations to terminate the number of epochs
+                k (int): Number of solutions to select in the selection function
+                p_crossover (float): probability of crossover
+                p_mutation (float): probability of mutation
+        Returns:
+            X_opt (list): Final Solution instance
+
+    """
+    
+    
+    
+    #ASSUMPTION 
+    #n >> 1
+    #N_pop >> 1
+    initial_population = create_Population(N_pop,n)
+   
+    current_solution = [] 
+    current_objective = 0
+    
+    best_objective = sys.maxsize #Set best objective value to infinity
+    best_solution = []
+    
+    
+    current_generation = []
+    epoch = 0 #Initial Iteration number
+    
+    
+    #Compute Objective for all solutions in iniital population
+    current_population = fitness(distance_data,flow_data,initial_population) #ISSUE OCCURS HERE initial_poulation does not change, needs to be current_population
```

Comment:
is the issue that occurs here resolved?

Reply:
yes had a bug previously. It is resolved now